core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1048,10 +1067,9 @@ class Log(@volatile var dir: File,
       lock synchronized {
         // remove the segments for lookups
         deletable.foreach(deleteSegment)
-        logStartOffset = math.max(logStartOffset, segments.firstEntry().getValue.baseOffset)
-        leaderEpochCache.clearAndFlushEarliest(logStartOffset)
-        producerStateManager.evictUnretainedProducers(logStartOffset)
-        updateFirstUnstableOffset()
+        val newLogStartOffset = math.max(logStartOffset, segments.firstEntry.getValue.baseOffset)
+        leaderEpochCache.clearAndFlushEarliest(newLogStartOffset)
```

Comment:
It seems that both line 1070 and 1071 can be moved into maybeIncrementLogStartOffset(). We can just call maybeIncrementLogStartOffset(segments.firstEntry.getValue.baseOffset) here. Inside maybeIncrementLogStartOffset, if startOffset changes, we do the leaderEpoch and producerState update.